### PR TITLE
[VarDumper] Add period caster

### PR DESCRIPTION
**Pull request by maidmaid**
| Q             | A
| ------------- | ---
| Branch?       | 3.4
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | https://github.com/symfony/symfony/pull/23591#discussion_r128307341
| License       | MIT
| Doc PR        | /

Result:

### CODE REVIEW

**Comment by maidmaid** [comment]
Proposal:

screenshot from 2017-07-25 20-48-58

**Comment by ro0NL** [comment]
If we only want start/end/interval to be nicely dumped, that would already work as is right?

Im not sure about the fancy period formatting :) though i like your creativity :+1:

**Comment by nicolas-grekas** [comment]
agree with @ro0NL, not really convinced myself...
maybe something like "every $period from $start to $end"?

**Comment by maidmaid** [comment]
Ok, I made tests green and I updated format. Example with end date (1) and with recurrences (2):

```
1) every + 1d, from 2017-01-01 00:00:00 (included) to 2017-01-03 00:00:00
2) every + 1d, from 2017-01-01 00:00:00 (included) for 2x
```

I kept tooltip with the dates included in the period. I find this useful, and you?

File: src/Symfony/Component/VarDumper/Caster/DateCaster.php [hunk 1]
```diff
@@ -79,4 +79,31 @@ public static function castTimeZone(\DateTimeZone $timeZone, array $a, Stub $stu
...
+    {
+        if (defined('HHVM_VERSION_ID') || \PHP_VERSION_ID < 50605) {
+            return $a;
+        }
+
+        $dates = array();
+        if (\PHP_VERSION_ID >= 70107) { // see https://bugs.php.net/bug.php?id=74639
+            $format = 'Y-m-d'.('000000' === $p->getStartDate()->format('His') && '000' === $p->getDateInterval()->format('%h%i%s') ? '' : ' H:i:s');
+            foreach (clone $p as $i => $d) {
+                $dates[] = sprintf('%s) %s', $i + 1, $d->format($format));
```
Comment (ro0NL) [inline src/Symfony/Component/VarDumper/Caster/DateCaster.php]: this list can become insanely long... not sure we should do it.

File: src/Symfony/Component/VarDumper/Caster/DateCaster.php [hunk 2]
```diff
@@ -79,4 +79,31 @@ public static function castTimeZone(\DateTimeZone $timeZone, array $a, Stub $stu
...
+            }
+        }
+
+        $period = sprintf(
+            'every %s, from %s (%s) %s',
+            self::formatInterval($p->getDateInterval()),
+            $p->getStartDate()->format('Y-m-d H:i:s'),
+            $p->include_start_date ? 'included' : 'excluded',
+            ($end = $p->getEndDate()) ? 'to '.$end->format('Y-m-d H:i:s') : 'for '.$p->recurrences.'x'
```
Comment (ro0NL) [inline src/Symfony/Component/VarDumper/Caster/DateCaster.php]: `, recurring x time/s`?

File: src/Symfony/Component/VarDumper/Caster/DateCaster.php (hunk 1 above)
Comment (maidmaid) [inline src/Symfony/Component/VarDumper/Caster/DateCaster.php]: Limit to *x* first items would be better? If yes, showing the first 12 items seem good to me.

Comment (ro0NL) [inline src/Symfony/Component/VarDumper/Caster/DateCaster.php]: dunno.. i would leave it out and keep the virtual period property.

Comment (maidmaid) [inline src/Symfony/Component/VarDumper/Caster/DateCaster.php]: Mmh, an overview of dates period can be very useful I think... Let's wait for other opinions ;)

File: src/Symfony/Component/VarDumper/Caster/DateCaster.php (hunk 2 above)
Comment (maidmaid) [inline src/Symfony/Component/VarDumper/Caster/DateCaster.php]: Updated

**Comment by maidmaid** [comment]
Fail not related.

File: src/Symfony/Component/VarDumper/Caster/DateCaster.php [hunk 3]
```diff
@@ -80,6 +80,36 @@ public static function castTimeZone(\DateTimeZone $timeZone, array $a, Stub $stu
...
+        if (defined('HHVM_VERSION_ID') || \PHP_VERSION_ID < 50605) {
+            return $a;
+        }
+
+        $dates = array();
+        if (\PHP_VERSION_ID >= 70107) { // see https://bugs.php.net/bug.php?id=74639
+            foreach (clone $p as $i => $d) {
+                $dates[] = sprintf('%s) %s', $i + 1, $d->format('Y-m-d H:i:s'));
+                if (11 === $i) {
+                    $dates[] = '...';
```
Comment (ro0NL) [inline src/Symfony/Component/VarDumper/Caster/DateCaster.php]: what about `x more` (see iterator_count)

File: src/Symfony/Component/VarDumper/Caster/DateCaster.php [hunk 4]
```diff
@@ -80,6 +80,36 @@ public static function castTimeZone(\DateTimeZone $timeZone, array $a, Stub $stu
...
+                }
+            }
+        }
+
+        $period = sprintf(
+            'every %s, from %s (%s) %s',
+            self::formatInterval($p->getDateInterval()),
+            $p->getStartDate()->format('Y-m-d H:i:s'),
+            $p->include_start_date ? 'included' : 'excluded',
+            ($end = $p->getEndDate()) ? 'to '.$end->format('Y-m-d H:i:s') : 'recurring '.$p->recurrences.' time/s'
```
Comment (ro0NL) [inline src/Symfony/Component/VarDumper/Caster/DateCaster.php]: maybe add `1 === $p->recurrences` case for time/s :)

**Review by ro0NL — APPROVED** [review]

File: src/Symfony/Component/VarDumper/Caster/DateCaster.php (hunk 3 above)
Comment (maidmaid) [inline src/Symfony/Component/VarDumper/Caster/DateCaster.php]: Yes, good idea. But ``iterator_count`` loops on all dates, what we want to avoid. We can find ``x`` with ``recurrences - limit`` in recurrences case and ``(end - start - limit) / interval`` in date case. WDYT?

File: src/Symfony/Component/VarDumper/Caster/DateCaster.php (hunk 4 above)
Comment (maidmaid) [inline src/Symfony/Component/VarDumper/Caster/DateCaster.php]: I don't see the interest of this. Always having a digital format is the simplest, no?

File: src/Symfony/Component/VarDumper/Caster/DateCaster.php [hunk 5]
```diff
@@ -80,6 +80,36 @@ public static function castTimeZone(\DateTimeZone $timeZone, array $a, Stub $stu
...
+        );
+
+        $p = array(Caster::PREFIX_VIRTUAL.'period' => new ConstStub($period, implode(PHP_EOL, $dates)));
```
Comment (nicolas-grekas) [inline src/Symfony/Component/VarDumper/Caster/DateCaster.php]: "\n" instead of PHP_EOL

Comment (maidmaid) [inline src/Symfony/Component/VarDumper/Caster/DateCaster.php]: Updated (Why exactly do you ask this change?)

File: src/Symfony/Component/VarDumper/Caster/DateCaster.php (hunk 3 above)
Comment (maidmaid) [inline src/Symfony/Component/VarDumper/Caster/DateCaster.php]: I added ``x more`` ;) I compared ``iterator_count()`` and my method with 1M of reccurences:
- 4160.1 ms with ``iterator_count()``
- 0.1 ms with my method

**Comment by maidmaid** [comment]
Can you review this PR please? :)

File: src/Symfony/Component/VarDumper/Caster/DateCaster.php (hunk 5 above)
Comment (robfrawley) [inline src/Symfony/Component/VarDumper/Caster/DateCaster.php]: Purely out of interest, I too would like to know why the change. Is it simply a Symfony convention or is there an another underlying reason for such?

**Comment by nicolas-grekas** [comment]
is the screenshot up to date?

**Comment by maidmaid** [comment]
Screenshot updated in PR description.

File: src/Symfony/Component/VarDumper/Caster/DateCaster.php [hunk 6]
```diff
@@ -20,6 +20,8 @@
  */
 class DateCaster
 {
+    const PERIOD_LIMIT = 12;
```
Comment (nicolas-grekas) [inline src/Symfony/Component/VarDumper/Caster/DateCaster.php]: why 12? I would have set 3 myself :)
the const should be removed (as we cannot make it private)

File: src/Symfony/Component/VarDumper/Caster/DateCaster.php [hunk 7]
```diff
@@ -80,6 +82,41 @@ public static function castTimeZone(\DateTimeZone $timeZone, array $a, Stub $stu
         return $filter & Caster::EXCLUDE_VERBOSE ? $z : $z + $a;
     }
 
+    public static function castPeriod(\DatePeriod $p, array $a, Stub $stub, $isNested, $filter)
+    {
+        if (defined('HHVM_VERSION_ID') || \PHP_VERSION_ID < 50605) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/VarDumper/Caster/DateCaster.php]: what happens before 5.6.5?

File: src/Symfony/Component/VarDumper/Caster/DateCaster.php [hunk 8]
```diff
@@ -80,6 +82,41 @@ public static function castTimeZone(\DateTimeZone $timeZone, array $a, Stub $stu
...
+            return $a;
+        }
+
+        $dates = array();
+        if (\PHP_VERSION_ID >= 70107) { // see https://bugs.php.net/bug.php?id=74639
+            foreach (clone $p as $i => $d) {
+                if (self::PERIOD_LIMIT === $i) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/VarDumper/Caster/DateCaster.php]: `if (3 === $i) {`

File: src/Symfony/Component/VarDumper/Caster/DateCaster.php [hunk 9]
```diff
@@ -80,6 +82,41 @@ public static function castTimeZone(\DateTimeZone $timeZone, array $a, Stub $stu
...
+                    // avoids iterator_count() function which can be expensive
+                    $now = new \DateTimeImmutable();
+                    $dates[] = sprintf('%s more', ($end = $p->getEndDate())
+                        ? ceil(($end->format('U.u') - $d->format('U.u')) / ($now->add($p->getDateInterval())->format('U.u') - $now->format('U.u')))
+                        : $p->recurrences - self::PERIOD_LIMIT
```
Comment (nicolas-grekas) [inline src/Symfony/Component/VarDumper/Caster/DateCaster.php]: `$p->recurrences - $i`

File: src/Symfony/Component/VarDumper/Caster/DateCaster.php [hunk 10]
```diff
@@ -80,6 +82,41 @@ public static function castTimeZone(\DateTimeZone $timeZone, array $a, Stub $stu
...
+    {
+        if (defined('HHVM_VERSION_ID') || \PHP_VERSION_ID < 50605) {
+            return $a;
+        }
+
+        $dates = array();
+        if (\PHP_VERSION_ID >= 70107) { // see https://bugs.php.net/bug.php?id=74639
+            foreach (clone $p as $i => $d) {
+                if (self::PERIOD_LIMIT === $i) {
+                    // avoids iterator_count() function which can be expensive
```
Comment (nicolas-grekas) [inline src/Symfony/Component/VarDumper/Caster/DateCaster.php]: not sure about the comment: nobody was wondering about iterator_count until read :)

File: src/Symfony/Component/VarDumper/Caster/DateCaster.php (hunk 6 above)
Comment (maidmaid) [reply src/Symfony/Component/VarDumper/Caster/DateCaster.php]: It's arbitrary but monthly periods can be displayed for a whole year with a limit of 12.

File: src/Symfony/Component/VarDumper/Caster/DateCaster.php (hunk 7 above)
Comment (maidmaid) [reply src/Symfony/Component/VarDumper/Caster/DateCaster.php]: ``getStartDate()``, ``getDateInterval()`` and ``getEndDate()`` methods were added in 5.6.5.

File: src/Symfony/Component/VarDumper/Caster/DateCaster.php (hunk 6 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/VarDumper/Caster/DateCaster.php]: 3 still looks enough to me, nobody will look at 12 values...

**Comment by maidmaid** [comment]
Changes done + screenshot updated.

**Review by nicolas-grekas — APPROVED** [review]

**Comment by nicolas-grekas** [comment]
Thank you @maidmaid.

**Comment by nicolas-grekas** [comment]
@maidmaid this is segfaulting on PHP 5.6, see https://travis-ci.org/symfony/symfony/jobs/269610022
Can you submit a fix please?

**Comment by maidmaid** [comment]
Ok, I check that now.
